Improve loading time when user logs in

FadhlanR opened the pull request:
In this PR, I implemented lazy loading for workspaces when the user logs in. This means the user doesn't need to wait for all workspace authentications to complete before seeing the workspace chooser or operator mode page. This works because on the workspace chooser page, we re-trigger authentication for each realm/workspace, and in the interact/code submode pages, we have a reauthentication mechanism that handles 401 errors when fetching cards or files. You can see in the screenshots below how workspaces and cards are now loaded lazily. The loading time should be faster on your end.

Login to workspace chooser:

https://github.com/user-attachments/assets/5da22f09-6bac-4936-8301-f0d12e0b80b9

Login and directly access submode:

https://github.com/user-attachments/assets/7388b614-d777-42ab-a002-4ab949b22639

FadhlanR committed:
DO NOT MERGE: checking login time

FadhlanR committed:
Check matrix client refresh routes

FadhlanR committed:
Fix matrix-client-start-total time

FadhlanR committed:
Fix matrix tests

FadhlanR committed:
Fix host test

FadhlanR committed:
Fix file tree test

FadhlanR committed:
Add comment and remove console time

FadhlanR committed:
Add test

FadhlanR committed:
Fix canonical url issue

FadhlanR committed:
Merge branch 'main' into cs-8574-improve-loading-time-when-login

FadhlanR committed:
Add comment

FadhlanR committed:
Merge branch 'main' into cs-8574-improve-loading-time-when-login

FadhlanR committed:
Update logic

jurgenwerk commented on packages/host/app/routes/index.ts:
```diff
@@ -73,8 +73,30 @@ export default class Index extends Route<void> {
       if (hostsOwnAssets) {
-        cardUrl = new URL(`/${cardPath}`, this.realm.defaultReadableRealm.path)
-          .href;
+        let realmUrl = Object.keys(this.realm.allRealmsInfo).find(
```
1. Is it guaranteed `allRealmsInfo` will be loaded at this point?
2. Would be good to have a test for this new card-realm matching behaviour
3. This model hook is getting huge - extract your addition into a function?

FadhlanR replied:
You're right — there's no guarantee that `allRealmsInfo` will be loaded here, so instead I use `availableActiveRealms` from `realmServerService`. We already have an existing test for the scenario where a user accesses Boxel using the canonical URL for a card that has an error. However, I added another test to ensure the new logic for card-realm matching works with URLs following the `[username]/[workspace name]` pattern. Please check it again.

jurgenwerk commented on packages/host/app/services/realm.ts:
```diff
@@ -641,7 +638,7 @@ export default class RealmService extends Service {
     { maxConcurrency: 1, enqueue: true },
     async (url: string): Promise<void> => {
-      if (this.knownRealm(url)) {
+      if (this.knownRealm(url, false)) {
```
Wondering why this was needed

FadhlanR replied:
We need this to avoid the error: "Attempted to update a tracked variable, but it had already been used previously in the same computation." Also, we don't need to use the tracked variable in this task.

jurgenwerk commented on packages/matrix/tests/login.spec.ts:
```diff
@@ -100,7 +100,58 @@ test.describe('Login', () => {
+    expect(boxelSessionBeforeCardLoaded).toBeNull();
+
+    await assertLoggedIn(page);
+    // the authentication to the realm could possibly processed when fetching card,
```
Same comment, I think some words are missing

FadhlanR replied:
Updated

jurgenwerk commented on packages/host/app/services/matrix-service.ts:
```diff
@@ -510,18 +510,24 @@ export default class MatrixService extends Service {
+
+        // Do not need to wait for theses to complete,
+        // in the workspace chooser we'll retrigger login and wait fo them to complete
+        // and when fetching cards or files we have reautentication mechanism.
```
There are some typos in this comment that are making it hard to follow

FadhlanR replied:
Updated

habdelra commented:
Also make sure to test this by going straight to an interact mode or code mode URL that has query param state when not logged in. the idea is that the login page shows for the user to authenticate, and after login, the operator-state-service will load the query param state. there might be some edge cases in this scenario where you are directly rendering a stack after login instead of just showing the dashboard.

FadhlanR committed:
Update comments and add test

FadhlanR commented:
> Also make sure to test this by going straight to an interact mode or code mode URL that has query param state when not logged in. the idea is that the login page shows for the user to authenticate, and after login, the operator-state-service will load the query param state. there might be some edge cases in this scenario where you are directly rendering a stack after login instead of just showing the dashboard.

@habdelra I added a screenshot of this scenario to the description of the PR. I also added the matrix tests.

FadhlanR committed:
Fix test

FadhlanR committed:
Merge branch 'main' into cs-8574-improve-loading-time-when-login

FadhlanR committed:
Deadlock handler

FadhlanR committed:
Merge branch 'main' into cs-8574-improve-loading-time-when-login
